pkg/controller/daemon_controller.go

```diff
@@ -0,0 +1,432 @@
+	// more pods until all the effects (expectations) of a daemon controller's create/delete have been observed.
+	dm.podStore.Store, dm.podController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dm.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return dm.kubeClient.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Pod{},
+		FullDaemonPodResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dm.addPod,
+			UpdateFunc: dm.updatePod, // TODO: add update function, and call it
+			DeleteFunc: dm.deletePod,
+		},
+	)
+	// Watch for new nodes or updates to nodes - daemons are launched on new nodes, and possibly when labels on nodes change,
+	dm.nodeStore.Store, dm.nodeController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dm.kubeClient.Nodes().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return dm.kubeClient.Nodes().Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Node{},
+		FullNodeResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dm.addNode,
+			UpdateFunc: dm.updateNode,
+			DeleteFunc: func(node interface{}) {},
+		},
+	)
+	dm.syncHandler = dm.syncDaemonController
+	return dm
+}
+
+func (dm *DaemonManager) Run(workers int, stopCh <-chan struct{}) {
+	go dm.dcController.Run(stopCh)
+	go dm.podController.Run(stopCh)
+	go dm.nodeController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go util.Until(dm.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down Daemon Controller Manager")
+	dm.queue.ShutDown()
+}
+
+func (dm *DaemonManager) worker() {
+	for {
+		func() {
+			key, quit := dm.queue.Get()
+			if quit {
+				return
+			}
+			defer dm.queue.Done(key)
+			err := dm.syncHandler(key.(string))
+			if err != nil {
+				glog.Errorf("Error syncing daemon controller: %v", err)
+			}
+		}()
+	}
+}
+
+func (dm *DaemonManager) enqueueAllControllers() {
+	daemonControllers, err := dm.dcStore.List()
+	if err != nil {
+		glog.Errorf("Error enqueueing daemon controllers: %v", err)
+		return
+	}
+	for i := range daemonControllers {
+		dm.enqueueController(&daemonControllers[i])
+	}
+}
+
+func (dm *DaemonManager) enqueueController(obj interface{}) {
+	key, err := controllerKeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	dm.queue.Add(key)
+}
+
+func (dm *DaemonManager) getPodDaemonController(pod *api.Pod) *api.DaemonController {
+	controllers, err := dm.dcStore.GetPodDaemonController(pod)
+	if err != nil {
+		glog.V(4).Infof("No controllers found for pod %v, daemon manager will avoid syncing", pod.Name)
```

Comment:
Yeah I'm on the fence about this one. It smells bad to use nil as a fence
in go, since we have multiple return paramters. What about returning
```(dc *api.DaemonController,
ok bool)```, which is also a go idiom?

It's also possible that you should not log this at all-- most pods will not
have a daemon controller associated. Or make it V(8) or something really
high. Also, if you do keep it, print the namespace as well as the name.

On Wed, Jul 8, 2015 at 8:35 PM, Ananya Kumar <notifications@github.com>
wrote:

> In pkg/controller/daemon_controller.go
> <https://github.com/vmarmol/kubernetes/pull/2#discussion_r34221279>:
>
> > +	}
> > +}
> > +
> > +func (dm *DaemonManager) enqueueController(obj interface{}) {
> > +	key, err := controllerKeyFunc(obj)
> > +	if err != nil {
> > +		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
> > +		return
> > +	}
> > +	dm.queue.Add(key)
> > +}
> > +
> > +func (dm *DaemonManager) getPodDaemonController(pod *api.Pod) *api.DaemonController {
> > +	controllers, err := dm.dcStore.GetPodDaemonController(pod)
> > +	if err != nil {
> > +		glog.V(4).Infof("No controllers found for pod %v, daemon manager will avoid syncing", pod.Name)
>
> I agreed with you initially, but in this case I think not returning an
> error avoids a lot of duplicate code. getPodDaemonController is called in 3
> places (addPod, updatePod, deletePod), and none of those methods are
> allowed to return an error. If you still feel I should return an error, I
> can log the error in addPod, updatePod, deletePod!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vmarmol/kubernetes/pull/2/files#r34221279>.
>